### PR TITLE
Implement secure horizontal scheme for federated learning

### PR DESCRIPTION
**Pull request by ZiyueXu77**
Implementation for the secure horizontal scheme (RFC https://github.com/dmlc/xgboost/issues/10170). 
Modifictions added beyond https://github.com/dmlc/xgboost/pull/10124
Following the same processor interface design with new functions to perform secure histogram aggregation (at external server).

### CODE REVIEW

File: src/collective/communicator.cc [hunk 1]
```diff
@@ -38,26 +64,46 @@ void Communicator::Init(Json const& config) {
...
+  std::string proc_params_map{};
+  plugin_name = OptionalArg<String>(config, "plugin_name", plugin_name);
+  // Initialize processor if plugin_name is provided
+  if (!plugin_name.empty()) {
+    std::map<std::string, std::string> loader_params = json_to_map(config, "loader_params");
+    std::map<std::string, std::string> proc_params = json_to_map(config, "proc_params");
+    processing::ProcessorLoader loader(loader_params);
+    processor_instance = loader.load(plugin_name);
+    processor_instance->Initialize(collective::GetRank() == 0, proc_params);
+  }
```
Comment (trivialfis) [inline src/collective/communicator.cc:82]: Do you have a document for the expected parameters?

File: src/collective/communicator.cc [hunk 2]
```diff
@@ -9,14 +10,39 @@
...
 thread_local CommunicatorType Communicator::type_{};
 thread_local std::string Communicator::nccl_path_{};
 
+std::map<std::string, std::string> json_to_map(xgboost::Json const& config, std::string key) {
+  auto json_map = xgboost::OptionalArg<xgboost::Object>(config, key, xgboost::JsonObject::Map{});
+  std::map<std::string, std::string> params{};
+  for (auto entry : json_map) {
+    std::string text;
+    xgboost::Value* value = &(entry.second.GetValue());
+    if (value->Type() == xgboost::Value::ValueKind::kString) {
```
Comment (trivialfis) [inline src/collective/communicator.cc:30]: `IsA<String>`

File: src/collective/aggregator.h [hunk 3]
```diff
@@ -96,8 +97,46 @@ void ApplyWithLabels(Context const*, MetaInfo const& info, HostDeviceVector<T>*
     }
     collective::Broadcast(&size, sizeof(std::size_t), 0);
 
-    result->Resize(size);
-    collective::Broadcast(result->HostPointer(), size * sizeof(T), 0);
+    if (info.IsSecure() && is_gpair) {
```
Comment (trivialfis) [inline src/collective/aggregator.h:100]: May I ask why the vertical federated learning section is being modified for horizontal learning?

File: tests/cpp/tree/hist/test_histogram.cc [hunk 4]
```diff
@@ -317,11 +316,6 @@ TEST(CPUHistogram, BuildHistDistColSplit) {
   RunWithInMemoryCommunicator(kWorkers, TestBuildHistogram, true, false, true, false);
 }
 
-TEST(CPUHistogram, BuildHistDistColSplitSecure) {
```
Comment (trivialfis) [inline tests/cpp/tree/hist/test_histogram.cc:320]: Why is this removed?

File: src/tree/hist/histogram.h [hunk 5]
```diff
@@ -180,47 +214,100 @@ class HistogramBuilder {
...
-          auto sibling_hist = this->hist_[sibling_nidx];
-          auto parent_hist = this->hist_[parent_id];
-          auto subtract_hist = this->hist_[subtraction_nidx];
-          common::SubtractionHist(subtract_hist, parent_hist, sibling_hist, r.begin(), r.end());
-        });
+        auto sibling_hist = this->hist_[sibling_nidx];
+        auto parent_hist = this->hist_[parent_id];
+        auto subtract_hist = this->hist_[subtraction_nidx];
+        common::SubtractionHist(subtract_hist, parent_hist, sibling_hist, r.begin(), r.end());
+      });
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:310]: Please don't make unnecessary changes.

File: src/processing/processor.h [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+    /*!
+     * \brief Handle buffers with encoded pairs received from broadcast
+     *
+     * \param size Output buffer size
+     * \param The encoded buffer
+     * \param The encoded buffer size
+     *
+     * \return The encoded buffer
+     */
+    virtual void* HandleGHPairs(size_t *size, void *buffer, size_t buf_size) = 0;
```
Comment (trivialfis) [inline src/processing/processor.h:61]: Can I translate this into: `Process -> Encrypt`, `Handle -> send/recv`?

File: src/processing/processor_loader.cc [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+namespace processing {
+    using LoadFunc = Processor *(const char *);
+
+    Processor* ProcessorLoader::load(const std::string& plugin_name) {
+        // Dummy processor for unit testing without loading a shared library
+        if (plugin_name == kMockProcessor) {
+            return new MockProcessor();
+        }
+
+        auto lib_name = "libproc_" + plugin_name;
```
Comment (trivialfis) [inline src/processing/processor_loader.cc:25]: Is it part of the protocol that the plugin must start with `libproc_`?

File: src/processing/processor_loader.cc [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        } else {
+            auto p = params[kLibraryPath];
+            if (p.back() != '/' && p.back() != '\\') {
+                p += '/';
+            }
+            lib_path = p + lib_file_name;
+        }
+
+#if defined(_WIN32) || defined(_WIN64)
+        handle_ = reinterpret_cast<void *>(LoadLibrary(lib_path.c_str()));
```
Comment (trivialfis) [inline src/processing/processor_loader.cc:50]: If we don't support windows, let's remove it.

File: src/processing/processor_loader.cc [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        if (!handle_) {
+            std::cerr << "Failed to load the dynamic library" << std::endl;
```
Comment (trivialfis) [inline src/processing/processor_loader.cc:52]: `LOG(FATAL)`

File: src/processing/processor_loader.cc [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            return NULL;
```
Comment (trivialfis) [inline src/processing/processor_loader.cc:53]: `nullptr`.

File: src/processing/processor_loader.cc [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+        }
+#endif
+
+        auto func = reinterpret_cast<LoadFunc *>(func_ptr);
+
+        return (*func)(plugin_name.c_str());
+    }
+
+    void ProcessorLoader::unload() {
```
Comment (trivialfis) [inline src/processing/processor_loader.cc:79]: Let's use C++ RAII instead.

File: src/tree/hist/evaluate_splits.h [hunk 12]
```diff
@@ -367,8 +367,9 @@ class HistEvaluator {
     auto const &cut_ptrs = cut.Ptrs();
     // Under secure vertical setting, only the active party is able to evaluate the split
     // based on global histogram. Other parties will receive the final best split information
-    // Hence the below computation is not performed by the passive parties
-    if ((!is_secure_) || (collective::GetRank() == 0)) {
+    // Hence the below computation is not performed by the passive parties for secure vertical
+    // All other cases need it
+    if (not (is_col_split_ && is_secure_ && collective::GetRank() != 0)) {
```
Comment (trivialfis) [inline src/tree/hist/evaluate_splits.h:372]: We should have a single function is collective to query this type of information.

File: src/tree/hist/histogram.h [hunk 13]
```diff
@@ -27,6 +28,7 @@
 #include "xgboost/logging.h"                    // for CHECK_GE
 #include "xgboost/span.h"                       // for Span
 #include "xgboost/tree_model.h"                 // for RegTree
+#include "../../processing/processor.h"         // for Processor
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:31]: Let's rename this into something more specific, `Processer` might not be clear to others.

File: src/tree/hist/histogram.h [hunk 14]
```diff
@@ -76,20 +79,51 @@ class HistogramBuilder {
...
-      bst_node_t const nidx = nodes_to_build[nid_in_set];
-      auto elem = row_set_collection[nidx];
-      auto start_of_row_set = std::min(r.begin(), elem.Size());
-      auto end_of_row_set = std::min(r.end(), elem.Size());
-      auto rid_set = common::RowSetCollection::Elem(elem.begin + start_of_row_set,
-                                                    elem.begin + end_of_row_set, nidx);
-      auto hist = buffer_.GetInitializedHist(tid, nid_in_set);
-      if (rid_set.Size() != 0) {
-        common::BuildHist<any_missing>(gpair_h, rid_set, gidx, hist, force_read_by_column);
+    if (is_distributed_ && is_col_split_ && is_secure_) {
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:82]: The title of this PR is horizontal, which is not column split?

File: src/tree/hist/histogram.h [hunk 15]
```diff
@@ -76,20 +79,51 @@ class HistogramBuilder {
...
+      // Call the interface to transmit gidx information to the secure worker
+      // for encrypted histogram compuation
+      auto slots = std::vector<int>();
+      auto num_rows = row_set_collection[0].Size();
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:86]: Do we need to use the first node `0` every time?

File: src/tree/hist/histogram.h [hunk 16]
```diff
@@ -180,47 +214,100 @@ class HistogramBuilder {
...
+        for (std::size_t hist_idx = 0; hist_idx < n; hist_idx++) {
+          reinterpret_cast<double *>(this->hist_[first_nidx].data())[hist_idx] =
+                  hist_aggr[hist_idx];
+        }
+      }
     }
 
     if (is_distributed_ && is_col_split_ && is_secure_) {
-      // Under secure vertical mode, we perform allgather for all nodes
+      // Under secure vertical mode, we perform allgather to get the global histogram.
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:260]: Let's focus on horizontal as described by the PR

File: src/processing/plugins/mock_processor.cc [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+    auto array_size = (*size_ptr - kPrefixLen)/8;
+    result.insert(result.end(), array_start, array_start + array_size);
+    rest_size -= *size_ptr;
+    ptr = ptr + *size_ptr;
+  }
+
+  return result;
+}
+
+void* MockProcessor::ProcessHistograms(std::size_t *size, const std::vector<double>& histograms) {
```
Comment (trivialfis) [inline src/processing/plugins/mock_processor.cc:117]: If this is for test only, let's move it out into the test module.

**Review by trivialfis — COMMENTED** [review]
Some questions in comments. I will open a new PR once the pipeline is cleared.

File: src/processing/processor.h [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+    void *handle_ = NULL;
+
+ public:
+    ProcessorLoader(): params{} {}
+
+    explicit ProcessorLoader(const std::map<std::string, std::string>& params): params(params) {}
+
+    Processor* load(const std::string& plugin_name);
+
+    void unload();
```
Comment (trivialfis) [inline src/processing/processor.h:127]: I couldn't find any reference to this unload method? Is it leaking the handle?

File: src/collective/aggregator.h (hunk 3 above)
Comment (ZiyueXu77) [reply src/collective/aggregator.h:100]: This one adds additional horizontal functions to the vertical_P2 PR, so it in fact includes everything that PR has, since that PR has not been merged, this one shows all the vertical modifications, 
"Modifictions added beyond https://github.com/dmlc/xgboost/pull/10124"

File: src/tree/hist/histogram.h (hunk 16 above)
Comment (ZiyueXu77) [reply src/tree/hist/histogram.h:260]: same as other vertical comments, after fixing and merging https://github.com/dmlc/xgboost/pull/10124, this PR should look much cleaner

File: src/processing/plugins/mock_processor.cc (hunk 17 above)
Comment (ZiyueXu77) [reply src/processing/plugins/mock_processor.cc:117]: make sense, mock serves for testing purpose @nvidianz what do you think?

File: src/processing/processor.h (hunk 18 above)
Comment (ZiyueXu77) [reply src/processing/processor.h:127]: @nvidianz

File: src/tree/hist/histogram.h (hunk 13 above)
Comment (ZiyueXu77) [reply src/tree/hist/histogram.h:31]: @nvidianz

File: src/tree/hist/histogram.h (hunk 15 above)
Comment (ZiyueXu77) [reply src/tree/hist/histogram.h:86]: good catch! I think we only need to perform InitAggregationContext once, but in this implementation we did it several times, I think we can check if the context has been initialized and only perform this step at the beginning. (@nvidianz is my understanding correct?)

File: tests/cpp/tree/hist/test_histogram.cc (hunk 4 above)
Comment (ZiyueXu77) [reply tests/cpp/tree/hist/test_histogram.cc:320]: It will not work and the tests implemented in TestBuildHistogram are not ready for secure tests - its logic is for non-secure version and does not cover the secure interface, etc. 
We need a test implementation compatible with secure schemes and add this (and RowSplitSecure) back

File: src/tree/hist/histogram.h (hunk 5 above)
Comment (ZiyueXu77) [reply src/tree/hist/histogram.h:317]: just to make the indent more consistent within this file

File: src/tree/hist/histogram.h (hunk 15 above)
Comment (ZiyueXu77) [reply src/tree/hist/histogram.h]: updated to make it happening only once

File: src/processing/plugins/mock_processor.cc [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+  int index = kPrefixLen;
+  for (auto value : pairs) {
+    for (std::size_t i = 0; i < 10; i++) {
+      memcpy(buf+index, &value, 8);
+      index += 8;
+    }
+  }
+
+  // Save pairs for future operations
+  this->gh_pairs_ = new std::vector<double>(pairs);
```
Comment (trivialfis) [inline src/processing/plugins/mock_processor.cc:36]: Is this vector freed?

File: src/tree/hist/evaluate_splits.h [hunk 20]
```diff
@@ -429,7 +430,7 @@ class HistEvaluator {
               all_entries[worker * entries.size() + nidx_in_set].split);
         }
       }
-      if (is_secure_) {
+      if (is_secure_ && is_col_split_) {
```
Comment (trivialfis) [inline src/tree/hist/evaluate_splits.h:433]: Looking at the code, the `is_col_split_` must be true.

Comment (ZiyueXu77) [reply src/tree/hist/evaluate_splits.h]: ah indeed, thanks! 
removed the redundant condition.

File: src/tree/hist/histogram.h [hunk 21]
```diff
@@ -180,47 +221,100 @@ class HistogramBuilder {
 
     common::BlockedSpace2d space(
         nodes_to_build.size(), [&](std::size_t) { return n_total_bins; }, 1024);
-    common::ParallelFor2d(space, this->n_threads_, [&](size_t node, common::Range1d r) {
+    common::ParallelFor2d(space, this->n_threads_, [&](std::size_t node, common::Range1d r) {
       // Merging histograms from each thread.
       this->buffer_.ReduceHist(node, r.begin(), r.end());
```
Comment (trivialfis) [inline src/tree/hist/histogram.h:226]: Is this reduce hist function call actually useful for secure training?

Comment (ZiyueXu77) [reply src/tree/hist/histogram.h:226]: I think so, this is the multi-thread part for each client

Comment (trivialfis) [reply src/tree/hist/histogram.h:226]: I thought the histogram is being built in the plugin? If so, any operation for the histogram in XGBoost should be invalid right?

Comment (ZiyueXu77) [reply src/tree/hist/histogram.h:226]: ah you are right! 
this reduce hist will have no impact for secure vertical training, it will be needed by secure horizontal, and non-secure federated schemes.
